service/history/workflow/cache/cache.go

```diff
@@ -372,3 +432,58 @@ func (c *CacheImpl) validateWorkflowID(
 
 	return nil
 }
+
+func (c *CacheImpl) planForNextEpoch() {
+	for {
+		time.Sleep(planningEpoch)
+		c.mut.Lock()
+		// Do not make any allocations if cache is not full.
+		if !c.cacheReachedLimit {
+			c.demand = make(map[string]int)
+			c.allocation = make(map[string]int)
+			c.allocation[unplannedCachePartition] = c.maxSize
+			c.mut.Unlock()
+			continue
+		}
+
+		c.allocation = make(map[string]int)
+		namespaces := make([]string, len(c.usage))
+		for ns := range c.usage {
+			namespaces = append(namespaces, ns)
+		}
+		sort.Slice(namespaces, func(i, j int) bool { return c.usage[namespaces[i]] < c.usage[namespaces[j]] })
+		capacity := c.maxSize - c.maxSize*unplannedCachePercentage/100
+		c.allocation[unplannedCachePartition] = c.maxSize * unplannedCachePercentage
+		for capacity > 0 {
+			count := 0
+			for _, ns := range namespaces {
+				if c.allocation[ns] < c.demand[ns] {
+					count++
```

Comment:
You need to deprecate capacity as you assign it and exit early if it hits 0

Reply:
I did not quite follow. Could you elaborate a little bit?